types/applications/dialog.d.ts

```diff
@@ -0,0 +1,328 @@
+/**
+ * Create a modal dialog window displaying a title, a message, and a set of buttons which trigger callback functions.
+ *
+ * @example <caption>Constructing a custom dialog instance</caption>
+ * ```typescript
+ * let d = new Dialog({
+ *  title: "Test Dialog",
+ *  content: "<p>You must choose either Option 1, or Option 2</p>",
+ *  buttons: {
+ *   one: {
+ *    icon: '<i class="fas fa-check"></i>',
+ *    label: "Option One",
+ *    callback: () => console.log("Chose One")
+ *   },
+ *   two: {
+ *    icon: '<i class="fas fa-times"></i>',
+ *    label: "Option Two",
+ *    callback: () => console.log("Chose Two")
+ *   }
+ *  },
+ *  default: "two",
+ *  render: html => console.log("Register interactivity in the rendered dialog"),
+ *  close: html => console.log("This always is logged no matter which option is chosen")
+ * });
+ * d.render(true);
+ * ```
+ */
+declare class Dialog extends Application {
+  /**
+   * @param data    - An object of dialog data which configures how the modal window is rendered
+   * @param options - Dialog rendering options, see {@link Application}
+   */
+  constructor(data: Dialog.Data, options: Partial<Dialog.Options>);
+
+  /* -------------------------------------------- */
+
+  /**
+   * @override
+   */
+  static get defaultOptions(): Dialog.Options;
+
+  /* -------------------------------------------- */
+
+  /**
+   * @override
+   */
+  static get title(): string;
+
+  /* -------------------------------------------- */
+
+  /**
+   * @override
+   */
+  getData(): { content: string; buttons: Record<string, DialogButton> };
+
+  /* -------------------------------------------- */
+
+  /**
+   * @override
+   */
+  activateListeners(html: JQuery): void;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle a left-mouse click on one of the dialog choice buttons
+   * @param event - The left-mouse click event
+   */
+  protected _onClickButton(event: MouseEvent): void;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle a keydown event while the dialog is active
+   * @param event - The keydown event
+   */
+  protected _onKeyDown(event: KeyboardEvent & { key: 'Escape' }): Promise<void>;
```

Comment:
The problem with this is that you don't always know at compile time what the key is. In particular, the type can be just `KeyboardEvent` at compile time while the `key` is `'Escape'` at runtime. With the current definition. we would then assume the return type to be `void`, i.e. the result to be `null` or `undefined`, while in reality a `Promise<void>` is returned. If you really want this overload, I think at least you need to make the return type of the second overload `void | Promise<void>`.

Reply:
I did not think of that, good point.